src/query_ctx.c

```diff
@@ -253,17 +253,37 @@ Graph *QueryCtx_GetGraph(void) {
 }
 
 // retrieve undo log
-UndoLog *QueryCtx_GetUndoLog(void) {
+UndoLog QueryCtx_GetUndoLog(void) {
+	QueryCtx *ctx = _QueryCtx_GetCtx();
+	ASSERT(ctx != NULL);
+	
+	if(ctx->undo_log == NULL) {
+		ctx->undo_log = UndoLog_New();
+	}
+	return ctx->undo_log;
+}
+
+// rollback the current command
+void QueryCtx_Rollback(void) {
 	QueryCtx *ctx = _QueryCtx_GetCtx();
-	ASSERT(ctx != NULL && ctx->undo_log != NULL);
+	ASSERT(ctx != NULL);
+
+	Graph_ResetReservedNode(ctx->gc->g);
+
+	if(ctx->undo_log == NULL) return;
 	
-	return &ctx->undo_log;
+	UndoLog_Rollback(ctx->undo_log);
+	ctx->undo_log = NULL;
```

Comment:
I think it would be nicer:
```c
UndoLog_Rollback(&ctx->undo_log);
```
Where `UndoLog_Rollback` free's and nullifies the input.

Reply:
you told me you don't want UndoLog_Rollback to get UndoLog* and that QueryCtx responsible to free itself